### PR TITLE
Feature/rdm 11402 optional document hash

### PR DESCRIPTION
**Pull request by sem-hmcts**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/RDM-11402

### Change description ###
Make document hash tokens optional

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/ccd/domain/service/createevent/CreateCaseEventService.java [hunk 1]
```diff
@@ -282,37 +264,55 @@ private CaseDetails saveCaseDetails(CaseDetails caseDetailsBefore, final CaseDet
...
-            caseDetails.setState(postState);
-        }
+    CaseDetails mergeUpdatedFieldsToCaseDetails(final Map<String, JsonNode> data,
+                                                final CaseDetails caseDetails,
+                                                final CaseEventDefinition caseEventDefinition,
+                                                final CaseTypeDefinition caseTypeDefinition) {
+
+        return Optional.ofNullable(data)
+            .map(x -> {
+                CaseDetails clonedCaseDetails = caseService.clone(caseDetails);
```
Comment (smathangi) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/createevent/CreateCaseEventService.java:285]: can we move this one step up as it's used in both map / orElse scenarios.

File: src/main/java/uk/gov/hmcts/ccd/domain/service/getcasedocument/CaseDocumentAttacher.java [hunk 2]
```diff
@@ -34,7 +34,7 @@
 import uk.gov.hmcts.ccd.endpoint.exceptions.ServiceException;
 import uk.gov.hmcts.ccd.v2.external.domain.DocumentHashToken;
 
-
+@Deprecated
 public class CaseDocumentAttacher {
```
Comment (smathangi) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/getcasedocument/CaseDocumentAttacher.java:38]: feel free to delete this class and related test class

File: src/main/java/uk/gov/hmcts/ccd/domain/service/createevent/CreateCaseEventService.java [hunk 3]
```diff
@@ -179,58 +148,69 @@ public CreateCaseEventResult createCaseEvent(String caseReference, CaseDataConte
...
-        String oldState = caseDetails.getState();
-        final CaseDetails savedCaseDetails =
-            saveCaseDetails(caseDetailsBefore,
-                caseDetails,
-                caseEventDefinition,
-                newState,
-                timeNow);
-        saveAuditEventForCaseDetails(aboutToSubmitCallbackResponse,
+        final List<DocumentHashToken> documentHashes = caseDocumentService.extractDocumentHashToken(
+            Optional.ofNullable(caseDetailsInDatabase.getData()).orElse(emptyMap()),
```
Comment (smathangi) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/createevent/CreateCaseEventService.java:177]: in the database `data` column is not null. so safe null checks may not be needed.

File: src/main/java/uk/gov/hmcts/ccd/domain/service/getcasedocument/CaseDocumentService.java [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        if (!applicationParams.isDocumentHashCheckingEnabled()) {
+            return;
+        }
+
+        final List<DocumentHashToken> violatingDocuments = caseDocumentUtils.getViolatingDocuments(
+            documentHashes
+        );
+
+        if (!violatingDocuments.isEmpty()) {
+            throw new ValidationException("Some message");  // TODO: suitable error message
```
Comment (smathangi) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/getcasedocument/CaseDocumentService.java:111]: Add error message like "Document hashTokens are missing for the documents: <document-ids>"

File: src/test/java/uk/gov/hmcts/ccd/domain/service/createevent/CreateCaseEventServiceTest.java [hunk 5]
```diff
@@ -342,20 +301,21 @@ void shouldUpdateUserDetailsWhenOnBehalfOfUserTokenIsPassed() {
             .withData(data)
             .withToken(TOKEN)
             .withIgnoreWarning(IGNORE_WARNING)
-            .withOnBehalfOfUserToken("Test_Token").build();
+            .withOnBehalfOfUserToken("Test_Token")
```
Comment (KleoG) [inline src/test/java/uk/gov/hmcts/ccd/domain/service/createevent/CreateCaseEventServiceTest.java:304]: Can maybe add test token as a field at the top as its used a couple of times

**Review by KleoG — APPROVED** [review]

**Review by sullivangareth — APPROVED** [review]
Clean and readable code.

Only minor comment/personal preference - I don't like generic names like `x` or `y`in lambda statements...but not a blocker

File: src/main/java/uk/gov/hmcts/ccd/domain/service/getcasedocument/CaseDocumentUtils.java [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+        try {
+            UUID.fromString(documentId);
+        } catch (RuntimeException e) {
+            throw new BadRequestException("The input document id is invalid");
+        }
+    }
+
+    public List<JsonNode> findDocumentNodes(@NonNull final Map<String, JsonNode> data) {
+        return data.values().stream()
+            .map(node -> node.findParents(DOCUMENT_URL))
```
Comment (smathangi) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/getcasedocument/CaseDocumentUtils.java:72]: Can you update this to use "DOCUMENT_BINARY_URL" instead please?

**Review by smathangi — APPROVED** [review]
Thank you @sem-hmcts  for all the good work.

**Pull request merged**